src/irmin/tree.ml

```diff
@@ -774,32 +774,62 @@ module Make (P : Private.S) = struct
         type acc.
         force:acc force ->
         uniq:uniq ->
-        pre:acc node_fn ->
-        post:acc node_fn ->
+        pre:acc node_fn option ->
+        post:acc node_fn option ->
         path:Path.t ->
         ?depth:depth ->
         node:(key -> _ -> acc -> acc Lwt.t) ->
         contents:(key -> contents -> acc -> acc Lwt.t) ->
+        tree:(key -> _ -> acc -> acc Lwt.t) option ->
         t ->
         acc ->
         acc Lwt.t =
-     fun ~force ~uniq ~pre ~post ~path ?depth ~node ~contents t acc ->
+     fun ~force ~uniq ~pre ~post ~path ?depth ~node ~contents ~tree t acc ->
       let marks =
         match uniq with
         | `False -> dummy_marks
         | `True -> empty_marks ()
         | `Marks n -> n
       in
+      let pre path bindings acc =
+        match pre with
+        | None -> Lwt.return acc
+        | Some pre ->
+            let s = Seq.fold_left (fun acc (s, _) -> s :: acc) [] bindings in
+            pre path s acc
+      in
+      let post path bindings acc =
+        match post with
+        | None -> Lwt.return acc
+        | Some post ->
+            let s = Seq.fold_left (fun acc (s, _) -> s :: acc) [] bindings in
+            post path s acc
+      in
       let rec aux : type r. (t, acc, r) folder =
        fun ~path acc d t k ->
-        let apply acc = node path t acc in
+        let apply acc =
+          match tree with
+          | Some f -> f path (`Node t) acc
+          | None -> node path t acc
+        in
         let next acc =
           match force with
-          | `True | `And_clear ->
-              (* XXX: Let's not call [to_map] when [Value] *)
-              let* m = to_map t >|= get_ok "fold" in
-              if force = `And_clear then clear ~depth:0 t;
-              (map [@tailcall]) ~path acc d (Some m) k
+          | `True | `And_clear -> (
+              match t.v with
+              | Map m ->
+                  if force = `And_clear then clear ~depth:0 t;
+                  (map [@tailcall]) ~path acc d (Some m) k
+              | Value (_, _, Some _) ->
+                  let* m = to_map t >|= get_ok "fold" in
```

Comment:
I would probably try to use `to_value` here too so that we never build a full map in memory when it's not necessary.

Reply:
Good idea, thanks!